### PR TITLE
Work on jsonapi.

### PR DESCRIPTION
**Pull request by stefanbutura**

### CODE REVIEW

**Comment by cristiroma** [comment]
@stefanbutura I don't see any test running on PR, can you check why?

**Comment by cristiroma** [comment]
There are some things to be added:

* Extend examples to include how to create taxonomy tags and paragraphs
* There must be a new role called "Legislation API" which allows manipulation of the Legislation entities - API users are required to have this role.
* Deny editing from the UI for the imported legislation (i.e. legislation created by the API user).

**Comment by cristiroma** [comment]
Done

File: docs/jsonapi/create_expression.json [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+        "data": [
+          {
+            "type": "paragraph--work_repeal",
+            "pid": "pid",
+            "id": "virtual",
+            "attributes": {
+              "field_repeal_date": "2019-08-23"
+            },
+            "relationships": {
+              "field_repeal_work": {
```
Comment (longhotsummer) [inline docs/jsonapi/create_expression.json]: Does this cause it to do a lookup for an existing work with the frbr_uri `/akn/za/2017/15/eng@2017-06-15`, and link that as the repealing work?

File: web/modules/custom/liiweb_api/src/Controller/LiiWebEntityResource.php [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+      }
+
+      $revision = $this->getRevisionFromFrbrUri($request->getRequestUri());
+      if (!empty($revision)) {
+        return $this->getResourceResponseError('Revision already exists.', 400);
+      }
+
+      /** @var \Drupal\Core\Entity\TranslatableInterface $revision */
+      $revision = $this->getRevisionWithPublicationDate($node, $date);
+      // A revision with that creation date does not exist, and the langcode is different from the original language - invalid request
```
Comment (longhotsummer) [inline web/modules/custom/liiweb_api/src/Controller/LiiWebEntityResource.php]:
```suggestion
      // A revision with that publication date does not exist, and the langcode is different from the original language - invalid request
```

File: web/modules/custom/liiweb_api/src/Controller/LiiWebEntityResource.php [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+      if (empty($revision) && $node->language()->getId() != $parsed_entity->language()->getId()) {
+        return $this->getResourceResponseError('Cannot create translations for revisions that do not exist. Please create a revision in the default language for that node with the requested date and try again.', 404);
```
Comment (longhotsummer) [inline web/modules/custom/liiweb_api/src/Controller/LiiWebEntityResource.php]: Does this mean we need to have an English revision before we can create another language? Can we change the default language on a per-site basis (won't be necessary in the short term, but will in the long term).

File: web/modules/custom/liiweb_api/src/Controller/LiiWebEntityResource.php [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+use Drupal\node\Entity\Node;
+use Drupal\node\NodeInterface;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpKernel\Exception\AccessDeniedHttpException;
+use Symfony\Component\HttpKernel\Exception\HttpException;
+use Symfony\Component\HttpKernel\Exception\NotFoundHttpException;
+
+class LiiWebEntityResource extends EntityResource {
+
+  const LANGCODE_MAPPING = [
```
Comment (longhotsummer) [inline web/modules/custom/liiweb_api/src/Controller/LiiWebEntityResource.php]: Is this used anywhere?

File: docs/jsonapi/create_expression.json (hunk 1 above)
Comment (longhotsummer) [reply docs/jsonapi/create_expression.json]: Looks like the answer is yes :)

**Comment by longhotsummer** [comment]
This looks great. I have suggested in Slack that it may be useful to store the work `frbr_uri` in addition to the expression `frbr_uri`. eg. store `/akn/za/act/2009/10` as well as `/akn/za/act/2009/10/eng@2019-01-02`.

**Comment by cristiroma** [comment]
Add example on how to post images and documents

File: web/modules/custom/liiweb_api/src/Controller/LiiWebEntityResource.php (hunk 3 above)
Comment (cristiroma) [reply web/modules/custom/liiweb_api/src/Controller/LiiWebEntityResource.php]: We've discussed this during the meeting. When a new node is created a language is selected for that node which becomes the "default" node language. When translations are created they are added to that expression, translations cannot exist without original version.

Yes, default language is per site.

**Pull request merged**